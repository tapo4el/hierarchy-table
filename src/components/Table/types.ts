import { OutputParametricSelector } from 'reselect';

import { AppState, Record } from '../../types';

export interface OwnProps {
    tableName: string;
    id?: string;
}
export interface TableProps {
    tableName: string;
    data: Record[];
    childList: string[];
}
export interface TableState {
    showChildrenFor: string[];
}

export type selectorType = OutputParametricSelector<AppState, string, Record[], (res: [AppState, string]) => Record[]>
