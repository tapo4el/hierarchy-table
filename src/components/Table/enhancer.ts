import { connect } from 'react-redux';

import { AppState, childTable } from '../../types';
import {
    getUsers,
    getPhones,
    getRelatives,
    getChildren,
} from '../../selectors';
import { OwnProps, TableProps } from './types';
import tableConfigs from '../../configs/tableConfigs';

const selectorsMap: any = {
    users: getUsers,
    relatives: getRelatives,
    phones: getPhones,
};

const childrenName = (tableName: string): childTable | undefined => tableConfigs[tableName].childTableName;

const mapStateToProps = (state: AppState, ownProps: OwnProps): TableProps => ({
    data: selectorsMap[ownProps.tableName](state, ownProps.id),
    tableName: ownProps.tableName,
    childList: childrenName(ownProps.tableName) ? getChildren(state, childrenName(ownProps.tableName)) : [],
});

export default connect(mapStateToProps, null);
