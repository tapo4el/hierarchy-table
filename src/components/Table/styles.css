body {
    margin: 10px;
}
.dataTable {
    width: 100%;
    border-top: 1px solid #dedede;
    border-left: 1px solid #dedede;
    color: #212121;
    font-family: Lato,'Helvetica Neue',Arial,Helvetica,sans-serif;
    font-size: 14px;
    border-collapse: separate;
}

.dataTable thead {
}
.dataTable th {
    background: rgb(220, 220, 220);
    font-weight: bold;
    padding: 10px 0 10px 0;
}
.dataTable td {
    padding: 10px;
    line-height: 17px;
    border-bottom: 1px solid rgba(34, 36, 38, 0.15);
    border-right: 1px solid rgba(34, 36, 38, 0.15);
}
.dataTable td.tableWrapper {
    padding: 20px 10px;
    background-color: #fff;
}

.dataTable tbody tr:nth-child(2n) {
    background-color: rgba(0,0,50,.02);
}
.dataTable tbody tr:hover {
    background-color: #fcfff5;
}
.tableWrapper>div {
    border: 1px solid rgba(34, 36, 38, 0.15);
    border-bottom: none;
    padding: 8px 0 8px 10px;
    font-weight: bold;
}
.dataTable td.arrow {
    width: 15px;
}
.arrow-down {
    width: 0;
    height: 0;
    border-left: 7px solid transparent;
    border-right: 7px solid transparent;

    border-top: 7px solid #212121;
}
.arrow-right {
    width: 0;
    height: 0;
    border-top: 7px solid transparent;
    border-bottom: 7px solid transparent;

    border-left: 7px solid #212121;
}
