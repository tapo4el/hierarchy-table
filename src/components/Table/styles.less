// default section
@default-border: 1px solid #dedede;
@default-font-size: 14px;
@default-padding: 10px;
@default-color: #212121;
// default section

body {
  margin: @default-padding;
}
.dataTable {
  width: 100%;
  border-top: @default-border;
  border-left: @default-border;
  color: @default-color;
  font-family: Lato, 'Helvetica Neue', Arial, Helvetica, sans-serif;
  font-size: @default-font-size;
  border-collapse: separate;

  th {
    background: rgb(220, 220, 220);
    font-weight: bold;
    padding: @default-padding 0;
  }

  td {
    padding: @default-padding;
    line-height: 17px;
    border-bottom: @default-border;
    border-right: @default-border;

    &.tableWrapper {
      padding: @default-padding*2 @default-padding;
      background-color: #fff;

      &>div {
        border: @default-border;
        padding: 8px 0 8px 10px;
        font-weight: bold;
      }
    }

    &.arrow {
      width: 15px;
      cursor: pointer;
    }
  }

  tbody {
    tr {
      &:nth-child(2n) {
        background-color: rgba(0,0,50,.02);
      }

      &:hover {
        background-color: #fcfff5;
      }
    }
  }
}
.arrowDown {
  width: 0;
  height: 0;
  border-left: 7px solid transparent;
  border-right: 7px solid transparent;
  border-top: 7px solid @default-color;
}
.arrowRight {
  width: 0;
  height: 0;
  border-top: 7px solid transparent;
  border-bottom: 7px solid transparent;
  border-left: 7px solid @default-color;
}
.removeButton {
  font-size: @default-font-size;
  cursor:pointer;
  border-radius: 5px;
}
