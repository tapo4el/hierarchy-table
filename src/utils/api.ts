import { UserResponse } from '../types';

export interface UserAPI {
    (): Promise<UserResponse[]>;
}

export default async function getUsers(): Promise<UserResponse[]> {
    const response = await fetch('/data', { method: 'GET' });
    return response.json();
}
