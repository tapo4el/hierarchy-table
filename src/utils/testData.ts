import {UserResponse, TableData, AppState, RemoveRow} from '../types';

export const response: UserResponse[] = [{
    data: {
        'Identification number': '34',
        Name: 'Jason',
        Gender: 'm',
        Risk: 'BITES',
        'Hair length': '1.6000000000',
        IQ: '91',
        'Admission date': 'Mon Feb 17 00:00:00 CET 1997',
        'Last breakdown': 'Wed Dec 03 03:09:55 CET 2014',
        'Yearly fee': '67932',
        'Knows the Joker?': 'false',
    },
    kids: {
        has_relatives: {// eslint-disable-line
            records: [
                {
                    data: {
                        'Relative ID': '1007',
                        'Patient ID': '34',
                        'Is alive?': 'true',
                        'Frequency of visits': '29',
                    },
                    kids: {
                        has_phone: {// eslint-disable-line
                            records: [
                                {
                                    data: {
                                        'Phone ID': '2008',
                                        'ID of the relative': '1007',
                                        Phone: '+(179)-982-0570',
                                    },
                                    kids: {

                                    },
                                },
                            ],
                        },
                    },
                },
            ],
        },
    },
}];

export const usersList = [{
    'Identification number': '34',
    Name: 'Jason',
    Gender: 'm',
    Risk: 'BITES',
    'Hair length': '1.6000000000',
    IQ: '91',
    'Admission date': 'Mon Feb 17 00:00:00 CET 1997',
    'Last breakdown': 'Wed Dec 03 03:09:55 CET 2014',
    'Yearly fee': '67932',
    'Knows the Joker?': 'false',
}];
export const relativesList = [{
    'Relative ID': '1007',
    'Patient ID': '34',
    'Is alive?': 'true',
    'Frequency of visits': '29',
}];
export const phonesList = [{
    'Phone ID': '2008',
    'ID of the relative': '1007',
    Phone: '+(179)-982-0570',
}];
export const finalAppState = {
    tableData: {
        users: usersList,
        relatives: {
            34: relativesList,
        },
        phones: {
            1007: phonesList,
        },
    },
};

export const initialAppState: AppState = {
    tableData: { users: [], relatives: {}, phones: {} },
};

export const initialState: TableData = { users: [], relatives: {}, phones: {} };

export const removeRowPayload: RemoveRow = {
    tableName: 'users',
    id: '34',
};
export const getCopy = (obj: any) => JSON.parse(JSON.stringify(obj));
