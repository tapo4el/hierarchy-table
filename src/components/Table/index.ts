import Table from './Table';
import enhancer from './enhancer';

export default enhancer(Table);
