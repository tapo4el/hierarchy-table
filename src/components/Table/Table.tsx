import * as React from 'react';

import './styles.less';

import { TableState } from './types';
import { MyProps } from './enhancer';
import tableConfigs from '../../configs/tableConfigs';
import { Record } from '../../types';
import ChildTable from './index';

class Table extends React.PureComponent<MyProps, TableState> {
    state: TableState = {
        showChildrenFor: [],
    };

    onClickHandler(id: string): void {
        const { showChildrenFor } = this.state;
        if (showChildrenFor.includes(id)) {
            this.setState({ showChildrenFor: showChildrenFor.filter(Id => Id !== id) });
        } else {
            this.setState({ showChildrenFor: showChildrenFor.concat(id) });
        }
    }

    renderRow(elem: Record): React.ReactNode {
        const {
            tableName,
            childList,
            removeRecord,
            parentId,
        } = this.props;
        const { showChildrenFor } = this.state;
        const { idField, childTableName, columns } = tableConfigs[tableName];
        const id = elem[idField];
        const hasChildTable = childList.includes(id);
        const isChildTableVisible = showChildrenFor.includes(id) && hasChildTable;

        return (
            <React.Fragment key={id}>
                <tr>
                    { childTableName && (
                        <td className="arrow" onClick={() => this.onClickHandler(id)} key="arrow" role="presentation">
                            { hasChildTable && <div className={isChildTableVisible ? 'arrowDown' : 'arrowRight'} /> }
                        </td>
                    )}
                    { Object.keys(elem).map(item => <td key={item}>{elem[item]}</td>) }
                    <td key="remove">
                        <button
                            type="button"
                            className="removeButton"
                            onClick={() => removeRecord({ parentId, tableName, id })}
                        >
                            Remove
                        </button>
                    </td>
                </tr>
                { isChildTableVisible && (
                    <tr>
                        <td colSpan={columns.length} className="tableWrapper">
                            <div>{`has_${childTableName}`}</div>
                            <ChildTable tableName={childTableName} id={id} />
                        </td>
                    </tr>
                ) }
            </React.Fragment>
        );
    }

    render(): React.ReactNode {
        const { data, tableName } = this.props;
        const { columns } = tableConfigs[tableName];

        return (
            <table className="dataTable">
                <thead>
                    <tr>
                        { columns.map(elem => <th key={elem.key}>{elem.title}</th>)}
                    </tr>
                </thead>
                <tbody>
                    { data.map(elem => this.renderRow(elem))}
                </tbody>
            </table>
        );
    }
}

export default Table;
