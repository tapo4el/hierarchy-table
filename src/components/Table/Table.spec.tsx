import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import Table from './index';
import { usersList, finalAppState } from '../../utils/testData';

const mockStore = configureStore();

describe('Table', () => {
    it('"users" should be rendered correctly if it has child table', () => {
        const store = mockStore(finalAppState);
        const component = renderer.create(
            // @ts-ignore
            <Table tableName="users" store={store} />,
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('"users" should be rendered correctly if it doesn`t have child table', () => {
        const usersWithoutChildren = {
            tableData: {
                users: usersList,
                relatives: {},
                phones: {},
            },
        };
        const store = mockStore(usersWithoutChildren);
        const component = renderer.create(
            // @ts-ignore
            <Table tableName="users" store={store} />,
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('"relatives" should be rendered correctly', () => {
        const store = mockStore(finalAppState);
        const component = renderer.create(
            // @ts-ignore
            <Table tableName="relatives" id="34" store={store} />,
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('"phones" should be rendered correctly', () => {
        const store = mockStore(finalAppState);
        const component = renderer.create(
            // @ts-ignore
            <Table tableName="phones" id="1007" store={store} />,
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('should show the child table after clicking on the parent row', async () => {
        const store = mockStore(finalAppState);
        const wrapper = mount(
            <Provider store={store}><Table tableName="users" /></Provider>,
        );
        expect(wrapper.find('.dataTable')).toHaveLength(1);
        await wrapper.find('td.arrow').at(0).simulate('click');
        expect(wrapper.find('.dataTable')).toHaveLength(2);
    });

    it('should hide the child table after clicking on the parent row second time', async () => {
        const store = mockStore(finalAppState);
        const wrapper = mount(
            <Provider store={store}><Table tableName="users" /></Provider>,
        );
        await wrapper.find('td.arrow').at(0).simulate('click');
        expect(wrapper.find('.dataTable')).toHaveLength(2);
        await wrapper.find('td.arrow').at(0).simulate('click');
        expect(wrapper.find('.dataTable')).toHaveLength(1);
    });

    it('should remove the table after clicking on the remove button', async () => {
        const store = mockStore(finalAppState);
        const wrapper = mount(
            <Provider store={store}><Table tableName="users" /></Provider>,
        );
        expect(wrapper.find('.dataTable')).toHaveLength(1);
        await wrapper.find('.removeButton').at(0).simulate('click');
        expect(store.getActions()).toMatchSnapshot();
    });

    it('should remove the child table after clicking on the child remove button', async () => {
        const store = mockStore(finalAppState);
        const wrapper = mount(
            <Provider store={store}><Table tableName="users" /></Provider>,
        );
        expect(wrapper.find('.dataTable')).toHaveLength(1);
        await wrapper.find('td.arrow').at(0).simulate('click');
        expect(wrapper.find('.dataTable')).toHaveLength(2);
        await wrapper.find('.removeButton').at(1).simulate('click');
        expect(store.getActions()).toMatchSnapshot();
    });
});
