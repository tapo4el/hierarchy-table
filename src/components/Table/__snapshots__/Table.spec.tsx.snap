// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table "phones" should be rendered correctly 1`] = `
<table
  className="dataTable"
>
  <thead>
    <tr>
      <th>
        Phone ID
      </th>
      <th>
        ID of the relative
      </th>
      <th>
        Phone
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      onClick={[Function]}
    >
      <td>
        2008
      </td>
      <td>
        1007
      </td>
      <td>
        +(179)-982-0570
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table "relatives" should be rendered correctly 1`] = `
<table
  className="dataTable"
>
  <thead>
    <tr>
      <th>
        
      </th>
      <th>
        Relative ID
      </th>
      <th>
        Patient ID
      </th>
      <th>
        Is alive?
      </th>
      <th>
        Frequency of visits
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      onClick={[Function]}
    >
      <td
        className="arrow"
      >
        <div
          className="arrow-right"
        />
      </td>
      <td>
        1007
      </td>
      <td>
        34
      </td>
      <td>
        true
      </td>
      <td>
        29
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table "users" should be rendered correctly if it doesn\`t have child table 1`] = `
<table
  className="dataTable"
>
  <thead>
    <tr>
      <th>
        
      </th>
      <th>
        Identification number
      </th>
      <th>
        Name
      </th>
      <th>
        Gender
      </th>
      <th>
        Risk
      </th>
      <th>
        Hair length
      </th>
      <th>
        IQ
      </th>
      <th>
        Admission date
      </th>
      <th>
        Last breakdown
      </th>
      <th>
        Yearly fee
      </th>
      <th>
        Knows the Joker?
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      onClick={[Function]}
    >
      <td
        className="arrow"
      />
      <td>
        34
      </td>
      <td>
        Jason
      </td>
      <td>
        m
      </td>
      <td>
        BITES
      </td>
      <td>
        1.6000000000
      </td>
      <td>
        91
      </td>
      <td>
        Mon Feb 17 00:00:00 CET 1997
      </td>
      <td>
        Wed Dec 03 03:09:55 CET 2014
      </td>
      <td>
        67932
      </td>
      <td>
        false
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table "users" should be rendered correctly if it has child table 1`] = `
<table
  className="dataTable"
>
  <thead>
    <tr>
      <th>
        
      </th>
      <th>
        Identification number
      </th>
      <th>
        Name
      </th>
      <th>
        Gender
      </th>
      <th>
        Risk
      </th>
      <th>
        Hair length
      </th>
      <th>
        IQ
      </th>
      <th>
        Admission date
      </th>
      <th>
        Last breakdown
      </th>
      <th>
        Yearly fee
      </th>
      <th>
        Knows the Joker?
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      onClick={[Function]}
    >
      <td
        className="arrow"
      >
        <div
          className="arrow-right"
        />
      </td>
      <td>
        34
      </td>
      <td>
        Jason
      </td>
      <td>
        m
      </td>
      <td>
        BITES
      </td>
      <td>
        1.6000000000
      </td>
      <td>
        91
      </td>
      <td>
        Mon Feb 17 00:00:00 CET 1997
      </td>
      <td>
        Wed Dec 03 03:09:55 CET 2014
      </td>
      <td>
        67932
      </td>
      <td>
        false
      </td>
    </tr>
  </tbody>
</table>
`;
